### PR TITLE
Add VXL image filters (color frequency, pixel-level averaging, image conversion)

### PR DESCRIPTION
**Pull request by mattdawkins**

### CODE REVIEW

File: sprokit/processes/core/frame_list_process.cxx [hunk 1]
```diff
@@ -78,6 +78,10 @@ create_config_trait( frame_time, double, "0.03333333", "Inter frame time in seco
                      "timestamps for sequential frames. This can be used to simulate a frame rate in a "
                      "video stream application.");
 
+create_config_trait( no_path_in_name, bool, "true",
```
Comment (linus-sherrill) [inline sprokit/processes/core/frame_list_process.cxx]: It is usually better to state these options in the positive, such "full_path_file_name" and then set to false. Remember the vidtk process disable option.

File: sprokit/processes/core/frame_list_process.cxx [hunk 2]
```diff
@@ -231,6 +237,16 @@ ::_step()
     ++d->m_frame_number;
     d->m_frame_time += d->m_config_frame_time;
 
+    // update filename
+    if ( d->m_no_path_in_name )
+    {
+      const size_t last_slash_idx = a_file.find_last_of("\\/");
```
Comment (linus-sherrill) [inline sprokit/processes/core/frame_list_process.cxx]: There might be a function in kwiver-sys that will return only the file name.

File: sprokit/processes/core/handle_descriptor_request_process.cxx [hunk 3]
```diff
@@ -128,50 +152,53 @@ ::_step()
...
 
-  // Get output matrix and detections
+  // Get output descriptors from internal pipeline
   vital::track_descriptor_set_sptr descriptors;
-  std::vector< vital::image_container_sptr > images;
 
-  vital::string_t filename;
-  vital::string_t stream_id;
+  // Get filepaths
+  boost::filesystem::path p( request->data_location() );
```
Comment (linus-sherrill) [inline sprokit/processes/core/handle_descriptor_request_process.cxx]: We are trying to remove boost usage in KWIVER. Is there an alternative using kwiver-sys to do this?

File: sprokit/processes/core/image_file_reader_process.cxx [hunk 4]
```diff
@@ -69,24 +69,27 @@ namespace algo = kwiver::vital::algo;
...
 
 create_config_trait( frame_time, double, "0.3333333", "Inter frame time in seconds. "
-                     "The generated timestamps will have the specified number of seconds in the generated "
-                     "timestamps for sequential frames. This can be used to simulate a frame rate in a "
-                     "video stream application.");
+  "The generated timestamps will have the specified number of seconds in the generated "
+  "timestamps for sequential frames. This can be used to simulate a frame rate in a "
+  "video stream application." );
+
+create_config_trait( no_path_in_name, bool, "true",
```
Comment (linus-sherrill) [inline sprokit/processes/core/image_file_reader_process.cxx]: Same comment as above about stating options in the positive.

File: vital/util/buffer.h [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+  /// \brief Check if there is an \a offset away from the last item.
+  ///
+  /// An \a offset of 0 refers to the last item.
+  ///
+  /// If <tt>has_datum_at(x)</tt> returns \c true, then
+  /// <tt>has_datum_at(y)</tt> will also return \c true for all 0 \<=
+  /// y \<= x.
+  virtual bool has_datum_at( unsigned offset ) const = 0;
+
+  virtual unsigned offset_of( Data const& ) const = 0;
```
Comment (linus-sherrill) [inline vital/util/buffer.h]: Missing documentation for method?

File: vital/util/buffer.h [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+class buffer
+{
+public:
+  virtual ~buffer() { }
+
+  /// The maximum number of elements that can be stored in the buffer.
+  virtual unsigned capacity() const = 0;
+
+  /// \brief The number of elements currently in the buffer.
+  virtual unsigned size() const = 0;
```
Comment (linus-sherrill) [inline vital/util/buffer.h]: Typically this would be a size_t for return type. Same for capacity().

File: vital/util/buffer.h (hunk 5 above)
Comment (russelldj) [reply vital/util/buffer.h:74]: This only implementation of this in `ring_buffer.h` which returns -1 and there it is documented as not implemented. @mattdawkins should this be removed or do you remember what the point was?

**Comment by russelldj** [comment]
Everything but the Sprokit changes should have made it into KWIVER in my branches. I'm not sure why those were there in the first place. It seems like this can get closed.

**Comment by mleotta** [comment]
If those sprokit changes are still needed by someone, they can cherry pick those changes into a new PR.  I will close this one.

**Pull request closed without merging**